### ISSUE TITLE
GLTF Serializer - Dynamic texture name is getting changed

### ISSUE THREAD

**Issue by VarshaKamble159** [BabylonJS/Babylon.js#4300]
Hi,

I am trying to add text to Babylon model.  Following is the code for the same.

```
var hotspotId = String(Date.now()) + Math.floor(Math.random() * 10000);
hotspotMesh = BABYLON.Mesh.CreatePlane(hotspotId, 25, this._scene, false);
            // outputplane.billboardMode = BABYLON.AbstractMesh.BILLBOARDMODE_ALL;
            hotspotMesh.material = new BABYLON.StandardMaterial("outputplane" + hotspotId, this._scene);
            // hotspotMesh.position = new BABYLON.Vector3(0, 0, 25);
            // hotspotMesh.scaling.y = 0.4;

            var outputplaneTexture = new BABYLON.DynamicTexture( hotspotId, 512, this._scene, true);
            hotspotMesh.material['diffuseTexture'] = outputplaneTexture;
            hotspotMesh.material['specularColor'] = new BABYLON.Color3(0, 0, 0);
            hotspotMesh.material['emissiveColor'] = new BABYLON.Color3(1, 1, 1);
            hotspotMesh.material.backFaceCulling = false;
            hotspotMesh.rotation = new BABYLON.Vector3(0, -3, 0);
            hotspotMesh.position.y = 2;
            hotspotMesh.position.z = 4;
            //outputplaneTexture.getContext().clearRect(0, 140, 512, 512);
            outputplaneTexture.drawText("Some Text", null, 140, "bold 80px verdana", "white", "transparent");
```

With this, text is getting added to model. After updating model, i am serializing it to GLTF. GLTF serializer returing serialized files. I am saving these files.
![gltferror](https://user-images.githubusercontent.com/9550268/39809278-f78713dc-539e-11e8-9146-15bb7230e4c9.png)

In this object, name of the dynamic texture is getting changed, due to which text plane is not getting added properly  after serializing.

Please find attached screenshots of model before and after serialization.
![before](https://user-images.githubusercontent.com/9550268/39809664-114283dc-53a0-11e8-94d1-0f28f22f6c0b.png)
![after](https://user-images.githubusercontent.com/9550268/39809663-110ec0c4-53a0-11e8-8034-965be9d79b2c.png)

Please suggest for the same.
Thanks.

Regards,
Varsha

**Comment by ravip** [comment]
Looks like DynamicTextures are not working for Babylon as well as serialized gltf.

**Comment by deltakosh** [comment]
This is correct. We can think about saving the texture to ping in a future release perhaps?

**Comment by kcoley** [comment]
Hi @VarshaKamble159 thanks for flagging!  I'll look into this.  I plan on having a fix for this by Friday.

**Comment by VarshaKamble159** [comment]
Hi @kcoley,
Thanks.
Looking forward to get this feature merged. :)

Regards,
Varsha

**Comment by kcoley** [comment]
Hi @VarshaKamble159 I made a recent update for dynamic textures which was merged today.  Feel free to give that a try and let me know if you run into any issues.

**Comment by VarshaKamble159** [comment]
Hi @kcoley,

I updated npm version of Babylonjs and babylonjs-serializer. My current versions are:
```
  "babylonjs": "^3.2.0",
   "babylonjs-serializers": "^3.2.0",
```
I am getting following error with this versions: 

` Property 'GLTF' is private and only accessible within class 'GLTF2Export'.`

This error is pointing to the code line
`  let returnData = SERIALIZERS.GLTF2Export.GLTF(this._scene, filename);`

here, I am serializing Babylon file to GLTF.

Please suggest for the same.
Thanks,
Varsha

**Comment by deltakosh** [comment]
Due to the nature of the serialized we had to change the export function to GLTFAsync
http://doc.babylonjs.com/extensions/gltfexporter#exporting-a-scene-to-gltf

**Comment by deltakosh** [comment]
@kcoley will get back to you ASAP regarding this issue

**Comment by kcoley** [comment]
Hi @VarshaKamble159,
To verify, are you wanting the name of the dynamic texture to be used during glTF export?  I made a change where the names of the textures were all unique strings in the case where dynamic textures all have the same name.  Though I can change this naming behavior so that it uses the assigned texture name, and if an identical name is present, append a number behind it.

As for .png files, glTF requires either .jpg or .png file extensions.  For export, I am using .png to handle cases with alpha textures.  If a dynamic texture is applied to a mesh, I assign the generated .png to the mesh.  

In short, I can change the export behavior to respect the name of the dynamic texture (with the appending of .png) and applied a number after the name if multiple dynamic textures share the same name.  Does that sound like that works for you?

Thanks,
Kacey

**Comment by VarshaKamble159** [comment]
Hi @kcoley,

So My use case is to add text dynamically to existing model and save this model. So the problem is when I am adding text, I am creating a texture with a particular name. It's just a name, with no extension.
`var outputplaneTexture = new BABYLON.DynamicTexture( hotspotId, 512, this._scene, true);`

While saving this, I get texture with names' texture_1.png', 'texture_2.png' and it gets saved.

When I am reloading this saved model, it throws an error as not found with texture name(i.e. hotspotId in this case).
So, I was curious about how to map texture with name hotspotId with saved(i.e. texture_1) file?

Also, for dynamic texture creation, we don't use any extension, as we are dynamically creating a texture. So how they get mapped with saved image files.

I hope it clears the issue.

Thanks,
Varsha

**Comment by kcoley** [comment]
Hi @VarshaKamble159 , thanks that clarifies things. 

I can make a change to respect the texture name that you set during the creation of the dynamic texture.  However, the core specification of glTF is not able to support dynamic textures, so during export, I have to convert the dynamic texture into a .png file, hence why the output has a file extension.  So in the case of your example above, the result would be ```hotspotId + '.png'```.  Therefore, you can map your file by appending the png extension to the name. you assign your dynamic texture.  Would this solution work for you?

I will start working on this change, and hope to have it completed by tomorrow.

Thanks,
~Kacey

**Comment by deltakosh** [comment]
Done

**Comment by VarshaKamble159** [comment]
Hi @kcoley ,

The serializer works perfectly for GLTF now. But I found another issue. This dynamic texture does not get saved with Babylon Serializer. Can you please help me with that?
Thanks.

Regards,
Varsha

**Comment by kcoley** [comment]
Hi @VarshaKamble159 , have you tried against Babylon version 3.3.0 alpha 11 ?  Or could you provide a sample playground where the dynamic texture doesn't export? 

This demo playground seems to work when exporting dynamic textures: https://www.babylonjs-playground.com/#5ZCGRM#27

**Comment by VarshaKamble159** [comment]
Hi @kcoley ,

Its getting serialized with GLTF, but it's not getting saved in .babylon structure. When we export it to gltf, a new file is getting created for this dynamic structure. The same should also be created for babylon serializers, right?
Correct me if I am wrong.
Thanks.

Regards,
Varsha

**Comment by deltakosh** [comment]
Correct. 
@kcoley: we need to add a serialize function to the dynamic texture:)

**Comment by kcoley** [comment]
Hi @VarshaKamble159 , sorry for the delay, but we added support for serializing dynamic textures.  Here is a sample playground: http://playground.babylonjs.com/#FU0ES5

**Comment by VarshaKamble159** [comment]
Hi @kcoley ,

Thanks for the updates. I tried the same. But still, I am not able to make it work. 
Attaching JSON file which contains data I am getting as the output of following code
```
let jsFile = BABYLON.SceneSerializer.Serialize(this._scene); //alert('ok')
            let js = JSON.stringify(jsFile);
console.log(' js content' , js);
```

I checked the serialized Babylon JSON, it is returning materials with proper IDs. Also, the mesh is having proper materialID. still, it's not working.
Please help me with this.
Attaching Babylon model, gltf model plus error screen for reference.
[cube.zip](https://github.com/BabylonJS/Babylon.js/files/2235170/cube.zip)
[cube_gltf.scnassets.zip](https://github.com/BabylonJS/Babylon.js/files/2235171/cube_gltf.scnassets.zip)

Error after loading serialized model :
![jsonerror](https://user-images.githubusercontent.com/9550268/43312100-ed716bf6-91a9-11e8-81ad-a8f3a1e22dcf.png)

JSON file which contains output of babylon serializer:
[JSON.txt](https://github.com/BabylonJS/Babylon.js/files/2235173/JSON.txt)

**Comment by VarshaKamble159** [comment]
Hi @kcoley ,

Any updates on this?

Thanks.
Regards,
Varsha

**Comment by kcoley** [comment]
Hi @VarshaKamble159 , sorry I missed your original reply.  I'll reopen this issue and look into it today.

**Comment by kcoley** [comment]
@VarshaKamble159 serializing currently works in the latest release of Babylon.
Here is a sample playground based on the reference code you posted: http://playground.babylonjs.com/#MD4LTD

The one thing you may need to add is a line to wait for the scene to be ready, since the procedural texture could still be generating:

```javascript
scene.whenReadyAsync().then(() => {
        let jsFile = BABYLON.SceneSerializer.Serialize(scene); //alert('ok')
        let js = JSON.stringify(jsFile);
        console.log(' js content' , js);
    });
```

Here is the generated Babylon.js scene file

[test.babylon.txt](https://github.com/BabylonJS/Babylon.js/files/2367612/test.babylon.txt)

Is this what you are expecting?

**Comment by deltakosh** [comment]
Seems like it is good. Closing for now

**Comment by VarshaKamble159** [comment]
Hi @kcoley ,

It worked perfectly. :+1: 

Thanks a lot.

Regards,
Varsha
